src/node_variables.jl

```diff
@@ -0,0 +1,541 @@
+    return _nv_integer_ref(nvref)
+end
+
+function _nv_integer_ref(nvref::NodeVariableRef)
+    backend = graph_backend(nvref.node)
+    ci = MOI.ConstraintIndex{MOI.VariableIndex,MOI.Integer}(
+        graph_index(backend, nvref).value
+    )
+    cref = JuMP.constraint_ref_with_index(backend, ci)
+    return cref
+end
+
+function JuMP.is_integer(nvref::NodeVariableRef)
+    return _moi_nv_is_integer(nvref)
+end
+
+function _moi_nv_is_integer(nvref::NodeVariableRef)
+    backend = graph_backend(nvref.node)
+    ci = MOI.ConstraintIndex{MOI.VariableIndex,MOI.Integer}(
+        graph_index(backend, nvref).value
+    )
+    return MOI.is_valid(backend, ci)
+end
+
+function JuMP.set_integer(nvref::NodeVariableRef)
+    node = JuMP.owner_model(nvref)
+    _set_dirty(node)
+    _moi_set_integer_nv(nvref)
+    return nothing
+end
+
+function _moi_set_integer_nv(nvref::NodeVariableRef)
+    node = JuMP.owner_model(nvref)
+    if _moi_nv_is_integer(nvref)
+        return nothing
+    elseif _moi_nv_is_binary(nvref)
+        error(
+            "Cannot set the variable_ref $(nvref) to integer as it " * "is already binary."
+        )
+    end
+    con = JuMP.ScalarConstraint(nvref, MOI.Integer())
+    _moi_add_node_constraint(node, con)
+    return nothing
+end
+
+function JuMP.unset_integer(nvref::NodeVariableRef)
+    JuMP.delete(JuMP.owner_model(nvref), JuMP.IntegerRef(nvref))
+    return nothing
+end
+
+### node variable binary
+
+function JuMP.BinaryRef(nvref::NodeVariableRef)
+    if !JuMP.is_binary(nvref)
+        error("Variable $nvref is not binary.")
+    end
+    return _nv_binary_ref(nvref)
+end
+
+function _nv_binary_ref(nvref::NodeVariableRef)
+    backend = graph_backend(nvref.node)
+    ci = MOI.ConstraintIndex{MOI.VariableIndex,MOI.ZeroOne}(
+        graph_index(backend, nvref).value
+    )
+    cref = JuMP.constraint_ref_with_index(backend, ci)
+    return cref
+end
+
+function JuMP.is_binary(nvref::NodeVariableRef)
+    return _moi_nv_is_binary(nvref)
+end
+
+function _moi_nv_is_binary(nvref::NodeVariableRef)
+    backend = graph_backend(nvref.node)
+    ci = MOI.ConstraintIndex{MOI.VariableIndex,MOI.ZeroOne}(
+        graph_index(backend, nvref).value
+    )
+    return MOI.is_valid(backend, ci)
+end
+
+function JuMP.set_binary(nvref::NodeVariableRef)
+    node = JuMP.owner_model(nvref)
+    _set_dirty(node)
+    _moi_set_binary_nv(nvref)
+    return nothing
+end
+
+function _moi_set_binary_nv(nvref::NodeVariableRef)
+    node = JuMP.owner_model(nvref)
+    if _moi_nv_is_binary(nvref)
+        return nothing
+    elseif _moi_nv_is_integer(nvref)
+        error(
+            "Cannot set the variable_ref $(nvref) to binary as it " * "is already integer."
+        )
+    end
+    con = JuMP.ScalarConstraint(nvref, MOI.ZeroOne())
+    _moi_add_node_constraint(node, con)
+    return nothing
+end
+
+function JuMP.unset_binary(nvref::NodeVariableRef)
+    JuMP.delete(JuMP.owner_model(nvref), JuMP.BinaryRef(nvref))
+    return nothing
+end
+
```

Comment:
You had asked if there was any other JuMP methods we might be missing. This is not essential, but having the `JuMP.relax_integrality` could be convenient. Maybe this is just a `# TODO:` for now, but it could be nice in the future. In the DDP solver, I do the integration manually.

Reply:
we also need `set_normalized_coefficient`. These may be simple copy-paste from JuMP.jl. Let's try to get these methods working in this PR.

Reply:
Sounds good. If you want me to try to add those, I can do that on Monday. If you want to do it yourself, that's fine too.

Reply:
I added the `relax_integrality` methods. Feel free to take a stab at `set_normalized_coefficient`. JuMP has a good [example implementation](https://github.com/jump-dev/JuMP.jl/blob/301d46e81cb66c74c6e22cd89fb89ced740f157b/src/variables.jl#L2721). You will want to call `MOI.Modify` on each possible backend like we do when adding [constraints](https://github.com/jalving/Plasmo.jl/blob/0bcbd891626266c8df8cae60980cdc2ad3cad652/src/optinode.jl#L182).